File: app/src/main/java/com/worldventures/dreamtrips/wallet/ui/settings/help/feedback/payment/model/PaymentFeedbackViewModel.java
package com.worldventures.dreamtrips.wallet.ui.settings.help.feedback.payment.model;

import android.databinding.BaseObservable;
import android.databinding.Bindable;

import com.worldventures.dreamtrips.BR;

public class PaymentFeedbackViewModel extends BaseObservable {

   private AttemptsViewModel attemptsView = new AttemptsViewModel();
   private MerchantViewModel merchantView = new MerchantViewModel();
   private PaymentTerminalViewModel terminalView = new PaymentTerminalViewModel();
   private AdditionalInfoViewModel infoView = new AdditionalInfoViewModel();
   private boolean canBeLost;

   @Bindable
   public AttemptsViewModel getAttemptsView() {
      return attemptsView;
   }

   public void setAttemptsView(AttemptsViewModel attemptsView) {
      this.attemptsView = attemptsView;
      notifyPropertyChanged(BR.attemptsView);
   }

   @Bindable
   public MerchantViewModel getMerchantView() {
      return merchantView;
   }

   public void setMerchantView(MerchantViewModel merchantView) {
      this.merchantView = merchantView;
      notifyPropertyChanged(BR.merchantView);
   }

   @Bindable
   public PaymentTerminalViewModel getTerminalView() {
      return terminalView;
   }

   public void setTerminalView(PaymentTerminalViewModel terminalView) {
      this.terminalView = terminalView;
      notifyPropertyChanged(BR.terminalView);
   }

   @Bindable
   public AdditionalInfoViewModel getInfoView() {
      return infoView;
   }

   public void setInfoView(AdditionalInfoViewModel infoView) {
      this.infoView = infoView;
      notifyPropertyChanged(BR.infoView);
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) {
         return true;
      }
      if (o == null || getClass() != o.getClass()) {
         return false;
      }

      PaymentFeedbackViewModel that = (PaymentFeedbackViewModel) o;

      if (attemptsView != null ? !attemptsView.equals(that.attemptsView) : that.attemptsView != null) {
         return false;
      }
      if (merchantView != null ? !merchantView.equals(that.merchantView) : that.merchantView != null) {
         return false;
      }
      if (terminalView != null ? !terminalView.equals(that.terminalView) : that.terminalView != null) {
         return false;
      }
      return infoView != null ? infoView.equals(that.infoView) : that.infoView == null;

   }

   @Override
   public int hashCode() {
      int result = attemptsView != null ? attemptsView.hashCode() : 0;
      result = 31 * result + (merchantView != null ? merchantView.hashCode() : 0);
      result = 31 * result + (terminalView != null ? terminalView.hashCode() : 0);
      result = 31 * result + (infoView != null ? infoView.hashCode() : 0);
      return result;
   }

   public void setCanBeLost(boolean canBeLost) {
      this.canBeLost = canBeLost;
   }

   public boolean isDataChanged() {
      return !canBeLost
            && (attemptsView.isDataChanged()
            || merchantView.isDataChanged()
            || terminalView.isDataChanged()
            || infoView.isDataChanged());
   }
}

File: wallet/src/main/java/com/worldventures/wallet/ui/settings/help/feedback/payment/model/PaymentFeedbackViewModel.kt
package com.worldventures.wallet.ui.settings.help.feedback.payment.model

import android.databinding.BaseObservable
import android.os.Parcel
import android.os.Parcelable

data class PaymentFeedbackViewModel(
      val attemptsView: AttemptsViewModel = AttemptsViewModel(),
      val merchantView: MerchantViewModel = MerchantViewModel(),
      val terminalView: PaymentTerminalViewModel = PaymentTerminalViewModel(),
      val infoView: AdditionalInfoViewModel = AdditionalInfoViewModel()
) : BaseObservable(), Parcelable {

   private var canBeLost: Boolean = false

   val isDataChanged: Boolean
      get() = !canBeLost && (this.attemptsView.isDataChanged
            || this.merchantView.isDataChanged
            || this.terminalView.isDataChanged
            || this.infoView.isDataChanged)

   constructor(parcel: Parcel) : this(
         parcel.readParcelable(AttemptsViewModel::class.java.classLoader),
         parcel.readParcelable(MerchantViewModel::class.java.classLoader),
         parcel.readParcelable(PaymentTerminalViewModel::class.java.classLoader),
         parcel.readParcelable(AdditionalInfoViewModel::class.java.classLoader)) {
      canBeLost = parcel.readByte() != 0.toByte()
   }

   fun setCanBeLost(canBeLost: Boolean) {
      this.canBeLost = canBeLost
   }

   override fun writeToParcel(parcel: Parcel, flags: Int) {
      parcel.writeParcelable(attemptsView, flags)
      parcel.writeParcelable(merchantView, flags)
      parcel.writeParcelable(terminalView, flags)
      parcel.writeParcelable(infoView, flags)
      parcel.writeByte(if (canBeLost) 1 else 0)
   }

   override fun describeContents(): Int = 0

   companion object CREATOR : Parcelable.Creator<PaymentFeedbackViewModel> {
      override fun createFromParcel(parcel: Parcel) = PaymentFeedbackViewModel(parcel)

      override fun newArray(size: Int): Array<PaymentFeedbackViewModel?> = arrayOfNulls(size)
   }
}
